sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -410,15 +417,22 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
 
     if (DDLUtils.isDatasourceTable(withStatsProps)) {
       val oldDef = client.getTable(db, withStatsProps.identifier.table)
-      // Sets the `schema`, `partitionColumnNames` and `bucketSpec` from the old table definition,
-      // to retain the spark specific format if it is. Also add old data source properties to table
-      // properties, to retain the data source table format.
-      val oldDataSourceProps = oldDef.properties.filter(_._1.startsWith(DATASOURCE_PREFIX))
+      // get the data source properties from old table definition, and add the new location entry.
+      val dataSourceProps = oldDef.properties.filter(_._1.startsWith(DATASOURCE_PREFIX)) ++
+        tableDefinition.storage.locationUri.map { location =>
+          DATASOURCE_LOCATION -> location
+        }
       val newDef = withStatsProps.copy(
+        // TODO: we may break the hive-compatibility format for location URI here, we should follow
+        // `createTable` and try to alter the table with `locationUri` set, if it's failed, then set
```

Comment:
I think this is based on the [codes in `createTable`](https://github.com/apache/spark/blob/d6eede9a36766e2d2294951b054d7557008a5662/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L313-L323)